File: packages/wasm/itest/helpers/faucet.js
'use strict'

// Minimal client for the Lightspark regtest faucet, mirroring the Rust
// harness (crates/breez-sdk/cli/tests/harness/faucet.rs).

const DEFAULT_URL = 'https://api.lightspark.com/graphql/spark/rc'
const MAX_RETRIES = 3

/**
 * Fund a regtest bitcoin address, returning the funding txid. Retries with
 * exponential backoff. Reads FAUCET_URL, FAUCET_USERNAME, and
 * FAUCET_PASSWORD from the environment.
 *
 * @param {string} address
 * @param {number} amountSats
 * @returns {Promise<string>}
 */
async function fundAddress(address, amountSats) {
  let lastError
  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    if (attempt > 0) {
      await new Promise((resolve) => setTimeout(resolve, 2 ** attempt * 1000))
    }
    try {
      return await tryFundAddress(address, amountSats)
    } catch (e) {
      lastError = e
    }
  }
  throw lastError
}

async function tryFundAddress(address, amountSats) {
  const url = process.env.FAUCET_URL || DEFAULT_URL
  const headers = { 'Content-Type': 'application/json' }
  const { FAUCET_USERNAME, FAUCET_PASSWORD } = process.env
  if (FAUCET_USERNAME && FAUCET_PASSWORD) {
    const credentials = Buffer.from(`${FAUCET_USERNAME}:${FAUCET_PASSWORD}`).toString('base64')
    headers.Authorization = `Basic ${credentials}`
  }

  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      operationName: 'RequestRegtestFunds',
      variables: { amount_sats: amountSats, address },
      query:
        'mutation RequestRegtestFunds($address: String!, $amount_sats: Long!) { ' +
        'request_regtest_funds(input: {address: $address, amount_sats: $amount_sats}) ' +
        '{ transaction_hash}}'
    })
  })
  const body = await response.text()
  if (!response.ok) {
    throw new Error(`faucet request failed with status ${response.status}: ${body}`)
  }
  const parsed = JSON.parse(body)
  if (parsed.errors && parsed.errors.length > 0) {
    throw new Error(`faucet returned errors: ${parsed.errors.map((e) => e.message).join(', ')}`)
  }
  if (!parsed.data) {
    throw new Error(`faucet response has no data: ${body}`)
  }
  return parsed.data.request_regtest_funds.transaction_hash
}

module.exports = { fundAddress }

File: crates/breez-sdk/cli/tests/harness/faucet.rs
//! Minimal client for the Lightspark regtest faucet, mirroring
//! `breez-itest/src/faucet.rs` (not a dev-dependency: that crate pulls
//! testcontainers and database drivers into the plain workspace test job).
//! Concurrency is bounded by the runner's `--test-threads` instead of a
//! semaphore.

use std::collections::HashMap;
use std::time::Duration;

use anyhow::{Context, Result, bail};
use platform_utils::{
    ContentType, DefaultHttpClient, HttpClient, add_basic_auth_header, add_content_type_header,
};
use serde::{Deserialize, Serialize};

const DEFAULT_URL: &str = "https://api.lightspark.com/graphql/spark/rc";
const MAX_RETRIES: u32 = 3;

#[derive(Serialize)]
struct GraphQlRequest {
    #[serde(rename = "operationName")]
    operation_name: String,
    variables: FaucetVariables,
    query: String,
}

#[derive(Serialize)]
struct FaucetVariables {
    amount_sats: u64,
    address: String,
}

#[derive(Deserialize)]
struct GraphQlResponse {
    data: Option<ResponseData>,
    errors: Option<Vec<GraphQlError>>,
}

#[derive(Deserialize)]
struct ResponseData {
    request_regtest_funds: RequestRegtestFunds,
}

#[derive(Deserialize)]
struct RequestRegtestFunds {
    transaction_hash: String,
}

#[derive(Deserialize)]
struct GraphQlError {
    message: String,
}

/// Fund a regtest bitcoin address, returning the funding txid. Retries with
/// exponential backoff. Reads `FAUCET_URL`, `FAUCET_USERNAME`, and
/// `FAUCET_PASSWORD` from the environment.
pub async fn fund_address(address: &str, amount_sats: u64) -> Result<String> {
    let mut last_error = None;
    for attempt in 0..=MAX_RETRIES {
        if attempt > 0 {
            tokio::time::sleep(Duration::from_secs(2u64.pow(attempt))).await;
        }
        match try_fund_address(address, amount_sats).await {
            Ok(txid) => return Ok(txid),
            Err(e) => last_error = Some(e),
        }
    }
    Err(last_error.expect("at least one attempt ran"))
}

async fn try_fund_address(address: &str, amount_sats: u64) -> Result<String> {
    let url = std::env::var("FAUCET_URL").unwrap_or_else(|_| DEFAULT_URL.to_string());
    let request = GraphQlRequest {
        operation_name: "RequestRegtestFunds".to_string(),
        variables: FaucetVariables {
            amount_sats,
            address: address.to_string(),
        },
        query: "mutation RequestRegtestFunds($address: String!, $amount_sats: Long!) { \
                request_regtest_funds(input: {address: $address, amount_sats: $amount_sats}) \
                { transaction_hash}}"
            .to_string(),
    };
    let body = serde_json::to_string(&request).context("failed to serialize faucet request")?;

    let mut headers = HashMap::new();
    add_content_type_header(&mut headers, ContentType::Json);
    if let (Ok(username), Ok(password)) = (
        std::env::var("FAUCET_USERNAME"),
        std::env::var("FAUCET_PASSWORD"),
    ) {
        add_basic_auth_header(&mut headers, &username, &password);
    }

    let response = DefaultHttpClient::default()
        .post(url, Some(headers), Some(body))
        .await
        .context("faucet request failed")?;
    if !response.is_success() {
        bail!(
            "faucet request failed with status {}: {}",
            response.status,
            response.body
        );
    }

    let parsed: GraphQlResponse = response
        .json()
        .with_context(|| format!("unparseable faucet response: {}", response.body))?;
    if let Some(errors) = parsed.errors {
        let messages: Vec<String> = errors.into_iter().map(|e| e.message).collect();
        bail!("faucet returned errors: {}", messages.join(", "));
    }
    let data = parsed.data.context("faucet response has no data")?;
    Ok(data.request_regtest_funds.transaction_hash)
}
